pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -172,23 +141,201 @@ func (d *MCPDiscovery) InstancesByPort(svc *model.Service, servicePort int, labe
 
 		for _, svcPort := range se.Ports {
 			for _, h := range se.Hosts {
-				if uint32(servicePort) == svcPort.Number {
+				if host.Name(h) == svc.Hostname && servicePort == int(svcPort.Number) {
 					// add not ready endpoint
 					out = append(out, d.notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
-					// add other endpoints
-					out = append(out, d.serviceInstances(se, svcPort, conf, h)...)
+				}
+			}
+		}
+	}
+	// add other endpoints
+	if svc != nil {
```

Comment:
This loop is not efficient, however I have created an issue which optimizes the discovery. Basically the idea is not to keep track of the notReadyEndpoints individually but rather extract them from already stored configs, which are already stored in a more flat data structure, which allows us to extract serviceInstances more efficiently.